rpc/src/v1/impls/eth.rs

```diff
@@ -143,9 +176,46 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM> EthClient<C, SN, S, M, EM> where
 		unwrap_provider(&self.accounts)
 	}
 
-	fn block(&self, id: BlockId, include_txs: bool) -> Result<Option<RichBlock>> {
+	fn rich_block(&self, id: BlockNumberOrId, include_txs: bool) -> Result<Option<RichBlock>> {
 		let client = &self.client;
-		match (client.block(id.clone()), client.block_total_difficulty(id)) {
+
+		let client_query = |id| (client.block(id), client.block_total_difficulty(id), client.block_extra_info(id));
+
+		let (block, difficulty, extra) = match id {
+			BlockNumberOrId::Number(BlockNumber::Pending) => {
+				let info = self.client.chain_info();
+				let pending_block = self.miner.pending_block(info.best_block_number);
+				let difficulty = { 
+					let latest_difficulty = self.client.block_total_difficulty(BlockId::Latest).expect("blocks in chain have details; qed");
+					let pending_difficulty = self.miner.pending_block_header(info.best_block_number).map(|header| *header.difficulty());
+
+				 	if let Some(difficulty) = pending_difficulty {
+						difficulty + latest_difficulty
+					} else {
+						latest_difficulty
+					}
+				};
+
+				let extra = pending_block.as_ref().map(|b| self.client.engine().extra_info(&b.header));
+
+				(pending_block.map(|b| encoded::Block::new(b.rlp_bytes(Seal::Without))), Some(difficulty), extra)
+			},
+
+			BlockNumberOrId::Number(num) => {
+				let id = match num {
+					BlockNumber::Latest => BlockId::Latest,
+					BlockNumber::Earliest => BlockId::Earliest,
+					BlockNumber::Num(n) => BlockId::Number(n),
```

Comment:
> In principle `BlockNumberOrId::Id(hash)` and `BlockNumberOrId::Number(BlockNumber::Num(num))` can refer to pending block as well.

Hm… I thought I've ported this part from the client code as is. But I agree, it seems reasonable. 

Also, I think, it would be better to abstract all those matches somehow to conform DRY principle. But the problem is that `Pending` handling is highly context dependent. I've tried to write a function that would match but that time I thought, only `Pending` should be handled by `Miner`. And it messed the logic in a way that verbose matches seemed the best option.

Anyway, I will look again for a way to handle it properly.